File: problem-098/solution.py
from typing import List, Set, Tuple


def find_word(board: List[List[str]], target: str) -> bool:
    if not board:
        return False

    row_length = len(board)
    col_length = len(board[0])

    for row in range(row_length):
        for col in range(col_length):
            visited: Set[Tuple[int, int]] = set()
            if search(board, row, col, target, 0, visited):
                return True


def search(board: List[List[str]], row: int, col: int, target: str, target_index: int, visited: Set[Tuple[int, int]]) -> bool:
    if not is_valid(board, row, col):
        return False
    if (row, col) in visited:
        return False
    if board[row][col] != target[target_index]:
        return False
    if target_index == len(target) - 1:
        return True

    visited.add((row, col))

    deltas = [
        (0, -1),
        (0, 1),
        (-1, 0),
        (1, 0)
    ]

    for delta in deltas:
        if search(board, row + delta[0], col + delta[1], target, target_index + 1, visited):
            return True

    visited.remove((row, col))
    return False


def is_valid(board: List[List[str]], row: int, col: int) -> bool:
    return row >= 0 and row < len(board) and col >= 0 and col < len(board[0])

test_grid = [
    ['A', 'B', 'C', 'E'],
    ['S', 'F', 'C', 'S'],
    ['A', 'D', 'E', 'E']
]

assert find_word(test_grid, 'ABCCED')
assert find_word(test_grid, 'SEE')
assert not find_word(test_grid, 'ABCB')

File: problem-098/solution.go
package problem098

type coordinate struct {
	x, y int
}

func findWord(board [][]byte, target string) bool {
	if len(board) == 0 {
		return false
	}

	rowLength := len(board)
	colLength := len(board[0])

	for i := 0; i < rowLength; i++ {
		for j := 0; j < colLength; j++ {
			visited := make(map[coordinate]bool)
			coor := coordinate{i, j}
			if search(board, coor, target, 0, visited) {
				return true
			}
		}
	}

	return false
}

func search(board [][]byte, coor coordinate, target string, targetIdx int, visited map[coordinate]bool) bool {
	if !isValid(board, coor) {
		return false
	}

	if visited[coor] {
		return false
	}

	if board[coor.x][coor.y] != target[targetIdx] {
		return false
	}

	if targetIdx == len(target)-1 {
		return true
	}

	visited[coor] = true

	deltas := []coordinate{
		{0, 1},
		{0, -1},
		{1, 0},
		{-1, 0},
	}

	for _, delta := range deltas {
		nextCoor := coordinate{
			x: coor.x + delta.x,
			y: coor.y + delta.y,
		}
		if search(board, nextCoor, target, targetIdx+1, visited) {
			return true
		}
	}

	delete(visited, coor)
	return false
}

func isValid(board [][]byte, coor coordinate) bool {
	return coor.x >= 0 && coor.x < len(board) && coor.y >= 0 && coor.y < len(board[0])
}
